src/Microsoft.AspNet.Mvc.Razor/TagHelpers/UrlResolutionTagHelper.cs

```diff
@@ -0,0 +1,212 @@
+                        if (TryResolveUrl(
+                            htmlStringValue,
+                            tryEncodeApplicationPath: true,
+                            resolvedUrl: out resolvedUrl))
+                        {
+                            attribute.Value = new HtmlString(resolvedUrl);
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool TryResolveUrl(string url, bool tryEncodeApplicationPath, out string resolvedUrl)
+        {
+            resolvedUrl = null;
+
+            if (url == null)
+            {
+                return false;
+            }
+
+            var urlLength = url.Length;
+
+            // Find the start of the potential application relative URL value.
+            var prefixEndIndex = -1;
+            while (++prefixEndIndex < urlLength && ValidAttributeWhitespaceChars.Contains(url[prefixEndIndex])) { }
+
+            // Before doing more work, ensure that the URL we're looking at is app relative.
+            if (prefixEndIndex < urlLength - 2 && url[prefixEndIndex] == '~' && url[prefixEndIndex + 1] == '/')
+            {
+                var valueEndIndex = prefixEndIndex - 1;
+                while (++valueEndIndex < urlLength && !ValidAttributeWhitespaceChars.Contains(url[valueEndIndex])) { }
+
+                var prefix = url.Substring(0, prefixEndIndex);
+                var urlValue = url.Substring(prefixEndIndex, valueEndIndex - prefixEndIndex);
+                var appRelativeUrl = UrlHelper.Content(urlValue);
+                var suffix = url.Substring(valueEndIndex);
+
+                if (tryEncodeApplicationPath)
+                {
+                    var postTildeSlashUrlValue = urlValue.Substring(2);
+
+                    if (!appRelativeUrl.EndsWith(postTildeSlashUrlValue, StringComparison.Ordinal))
+                    {
+                        throw new InvalidOperationException(
+                            Resources.FormatCouldNotResolveApplicationRelativeUrl_TagHelper(
+                                url,
+                                nameof(IUrlHelper),
+                                nameof(IUrlHelper.Content),
+                                "removeTagHelper",
+                                typeof(UrlResolutionTagHelper).FullName,
+                                typeof(UrlResolutionTagHelper).GetTypeInfo().Assembly.GetName().Name));
+                    }
+
+                    var applicationPath = appRelativeUrl.Substring(0, appRelativeUrl.Length - postTildeSlashUrlValue.Length);
+                    var encodedApplicationPath = HtmlEncoder.HtmlEncode(applicationPath);
+
+                    resolvedUrl = string.Concat(prefix, encodedApplicationPath, postTildeSlashUrlValue, suffix);
+                }
+                else
+                {
+                    resolvedUrl = string.Concat(prefix, appRelativeUrl, suffix);
```

Comment:
It's doubtful we need to go make every place in Razor consistent; I just wonder whether all this extra new code is worthwhile.